tendermint/src/tendermint.rs

```diff
@@ -390,52 +391,104 @@ impl<TProtocol: Protocol> Tendermint<TProtocol> {
         cx: &mut Context<'_>,
         should_export_state: &mut bool,
     ) -> Option<TaggedAggregationMessage<TProtocol::AggregationMessage>> {
-        // the state changes if there is at least one message to process as they are kept for future reference.
-        while let Poll::Ready(Some(mut message)) = self.level_update_stream.poll_next_unpin(cx) {
-            // signatures!
-            let id = message.tag;
-            if id.0 > self.state.current_round {
-                // future contributions need to be tracked separately, as once a round has f+1 contributions
-                // it can be fast tracked
-                let contributors = self
-                    .future_contributions
-                    .entry(id.0)
-                    .and_modify(|bitset| {
-                        *bitset |= message.aggregation.all_contributors();
-                    })
-                    .or_insert_with(|| message.aggregation.all_contributors());
-
-                // check if the skip ahead condition is fulfilled. If so, the state machine can be set to propose for round id.0
-                if contributors.len() >= TProtocol::F_PLUS_ONE {
-                    // skip to that round
-                    *should_export_state = true;
-                    self.state.current_round = id.0;
-                    self.state.current_step = Step::Propose;
-                    self.future_contributions
-                        .retain(|round, _contributors| round > &self.state.current_round);
-
-                    // Return to account the state change
-                    return None;
+        // Has any of the polled futures done something? If so, don't return. If we're out of
+        // futures to poll, return.
+        let mut did_something = true;
+        while did_something {
+            did_something = false;
+
+            // First, check if the future round aggregation verification has completed.
+            if let Some(verification) = &mut self.future_round_verification {
+                if let Poll::Ready(verification_result) = verification.poll_unpin(cx) {
+                    did_something = true;
+                    self.future_round_verification = None;
+
+                    // If the signature was found to be valid, add the contributing validators to
+                    // the set of contributors for that future round.
+                    if let Ok((round, new_contributors)) = verification_result {
+                        if round > self.state.current_round {
+                            // future contributions need to be tracked separately, as once a round has f+1 contributions
+                            // it can be fast tracked
+                            let contributors = self.future_contributions.entry(round).or_default();
+                            *contributors |= new_contributors;
```

Comment:
Maybe consider sticking with the Entry API here. 
```suggestion
                            self.future_contributions.entry(round)
                                .and_modify(|contributors| *contributors |= new_contributors)
                                .or_insert(new_contributors);
```

Reply:
Same ownership problem, but here, I also actually like the current version better.